pkg/operator/controller/gatewayapi/controller.go

```diff
@@ -0,0 +1,136 @@
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DependentControllers is a list of controllers that watch Gateway API
+	// resources.  The gatewayapi controller starts these controllers once
+	// the Gateway API CRDs have been created.
+	DependentControllers []controller.Controller
+}
+
+// reconciler reconciles gatewayclasses.
+type reconciler struct {
+	config Config
+
+	client           client.Client
+	cache            cache.Cache
+	recorder         record.EventRecorder
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and creates or
+// reconciles the Gateway API CRDs.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	var fg configv1.FeatureGate
+	if err := r.cache.Get(ctx, request.NamespacedName, &fg); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("featuregate not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if !featureIsEnabled("GatewayAPI", &fg) {
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.ensureGatewayAPICRDs(ctx); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.startControllers.Do(func() {
+		for i := range r.config.DependentControllers {
+			c := &r.config.DependentControllers[i]
+			go func() {
+				if err := (*c).Start(ctx); err != nil {
+					log.Error(err, "cannot start controller")
+				}
+			}()
+		}
+	})
+
+	return reconcile.Result{}, nil
+}
+
+// featureIsEnabled takes a feature name and a featuregate config API object and
+// returns a Boolean indicating whether the named feature is enabled.
+//
+// This function determines whether a named feature is enabled as follows:
+//
+//   - First, if the featuregate's spec.featureGateSelection.featureSet field is
+//     set to "CustomNoUpgrade", then the feature is enabled if, and only if, it
+//     is specified in spec.featureGateSelection.customNoUpgrade.enabled.
+//
+//   - Second, if spec.featureGateSelection.featureSet is set to a value that
+//     isn't defined in configv1.FeatureSets, then the feature is *not* enabled.
+//
+//   - Finally, the feature is enabled if, and only if, the feature is specified
+//     in configv1.FeatureSets[spec.featureGateSelection.featureSet].enabled.
+func featureIsEnabled(feature string, fg *configv1.FeatureGate) bool {
+	if fg.Spec.FeatureSet == configv1.CustomNoUpgrade {
+		if fg.Spec.FeatureGateSelection.CustomNoUpgrade == nil {
+			return false
+		}
+		for _, f := range fg.Spec.FeatureGateSelection.CustomNoUpgrade.Enabled {
+			if f == feature {
+				return true
+			}
+		}
+		return false
+	}
+
+	if fs, ok := configv1.FeatureSets[fg.Spec.FeatureSet]; ok {
```

Comment:
This is just extra, right?

Reply:
This is in case a feature set (such as "TechPreviewNoUpgrade") includes the "GatewayAPI" feature.  It isn't needed at this time because there is no feature set that includes the "GatewayAPI" feature; it's just here for correctness and future-proofing.